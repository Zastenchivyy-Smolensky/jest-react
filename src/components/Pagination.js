import React, { useState } from "react";

export const Pagination = ({ max }) => {
  const [currentIndex, setCurrentIndex] = useState(1);
  const pageIndexArray = [...Array(max)].map((_, index) => index + 1);

  return (
    <div>
      <ul>
        {pageIndexArray.map((index) => (
          <li key={`paginationItem_${index}`}>
            {index} {index === currentIndex && <span>◀︎ now</span>}
          </li>
        ))}
      </ul>
      <div>
        <button
          onClick={() => setCurrentIndex((prev) => prev - 1)}
          disabled={currentIndex <= 1}
        >
          prev
        </button>
        <button
          onClick={() => setCurrentIndex((prev) => prev + 1)}
          disabled={currentIndex >= max}
        >
          next
        </button>
      </div>
    </div>
  );
};
