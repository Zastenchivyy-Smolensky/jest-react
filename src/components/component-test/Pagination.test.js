import { render, screen, fireEvent } from "@testing-library/react";
import { Pagination } from "../Pagination";

describe("Pagination", () => {
  test('currentIndexとindexが一致した時に、"◀︎ now"が表示される', () => {
    // 対象Componentのレンダリング
    render(<Pagination max={5} />);
    expect(screen.queryByText(/1/)).toHaveTextContent(/◀︎ now/);
    expect(screen.queryByText(/2/)).not.toHaveTextContent(/◀︎ now/);
    // ユーザーの操作
    fireEvent.click(screen.queryByText(/next/));
    expect(screen.queryByText(/2/)).toHaveTextContent(/◀︎ now/);
    expect(screen.queryByText(/1/)).not.toHaveTextContent(/◀︎ now/);
  });

  test("currentIndexが1以下の時、prevボタンがdisabledになる", () => {
    render(<Pagination max={5} />);
    expect(screen.queryByText(/prev/)).toBeDisabled();
    fireEvent.click(screen.queryByText(/next/));
    expect(screen.queryByText(/prev/)).toBeEnabled();
    fireEvent.click(screen.queryByText(/prev/));
    expect(screen.queryByText(/prev/)).toBeDisabled();
  });

  test("currentIndexがmax以上の時、nextボタンがdisabledになる", () => {
    render(<Pagination max={5} />);
    fireEvent.click(screen.queryByText(/next/));
    fireEvent.click(screen.queryByText(/next/));
    fireEvent.click(screen.queryByText(/next/));
    fireEvent.click(screen.queryByText(/next/));
    expect(screen.queryByText(/next/)).toBeDisabled();
  });

  test("snapshot", () => {
    const { asFragment } = render(<Pagination max={5} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
