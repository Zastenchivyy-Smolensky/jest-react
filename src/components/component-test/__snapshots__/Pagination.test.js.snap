// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination snapshot 1`] = `
<DocumentFragment>
  <div>
    <ul>
      <li>
        1 
        <span>
          ◀︎ now
        </span>
      </li>
      <li>
        2 
      </li>
      <li>
        3 
      </li>
      <li>
        4 
      </li>
      <li>
        5 
      </li>
    </ul>
    <div>
      <button
        disabled=""
      >
        prev
      </button>
      <button>
        next
      </button>
    </div>
  </div>
</DocumentFragment>
`;
